library/core/src/intrinsics.rs

```diff
@@ -2709,6 +2709,9 @@ pub const unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: us
 ///
 /// * `dst` must be [valid] for writes of `count * size_of::<T>()` bytes.
 ///
+/// * `src` must remain valid for reads even after `dst` is written, and vice versa.
+///   (In other words, there cannot be aliasing restrictions on the use of these pointers.)
+///
```

Comment:
I didn't quite understand what this meant until I checked out #81005.

Maybe there's a different way of wording this?

Would it work to not add this bullet point, but instead add "during the entire call to swap()" to the first two bullet points? (Maybe with a clarifying note saying that this means that the src and dst regions may not overlap.)

Reply:
They may overlap though, the pointers just must allow overlapping access. So for instance passing `ptr` and `ptr.add(N)` is fine even if things overlap. However passing `&mut *ptr` twice generates two fresh mutable references and now if things overlap, that lead to UB due to aliasing violations.

It's hard to precisely say this while the aliasing model is still being determined.

> add "during the entire call to swap()" to the first two bullet points?

If we do that I think we have to say what happens during the call. So we could say something like

* `src` must be [valid] for reads of `count * size_of::<T>()` bytes, and must remain valid even if `dst` is written to for `count * size_of::<T>()` bytes.

But that seems very similar to the current wording, so I am not sure if it is clear enough.